### DISCUSSION TITLE
Use of icons to indicate the status of elements, properties, and APIs

### DISCUSSION THREAD

**Discussion by schalkneethling** [mdn/mdn-community#69] [Platform]
In various areas, such as the sidebars, main content, and browser compatibility tables, we use icons as a means to indicate whether an element, property, or API is:

- Obsolete
- Experimental
- Non-standard
- Deprecated
- and possibly other statuses

Currently, this is used inconsistently in terms of the use of color; for example:

## In the sidebar

![Screenshot icons used in the sidebar. In this instance, the trash can is red and the exclamation mark blue](https://user-images.githubusercontent.com/10350960/167730549-9d58d24f-17cb-4ea2-a80f-b9c81bdd4818.png)

## In the main content

![Screenshot of icons used in the main content. Here the trash can icon is a dark gray.](https://user-images.githubusercontent.com/10350960/167730637-1e5a54de-52f7-4fc3-8237-6d406a8d93aa.png)

## In BCD tables

![Screenshot of icons used in BCD tables. Here the icons are a light gray.](https://user-images.githubusercontent.com/10350960/167730776-13209017-8580-437f-9ae5-b50ad8ffe23e.png)

These icons have created constant problems with the layout of the items in the sidebars that have yet to be solved.

- https://github.com/mdn/yari/issues/1721
- https://github.com/mdn/yari/issues/2221
- https://github.com/mdn/yari/discussions/4033
- https://github.com/mdn/content/discussions/7535#discussioncomment-2710838
- https://github.com/mdn/yari/issues/2221
- https://github.com/mdn/yari/issues/5571

These icons are also not exposed in an accessible manner, especially in the sidebars, for example, https://github.com/mdn/yari/issues/5905

## Kumascript macros

Several Kumascript macros exist solely to display these icons.

- https://github.com/mdn/yari/blob/main/kumascript/macros/Non-standardGeneric.ejs
- https://github.com/mdn/yari/blob/main/kumascript/macros/DeprecatedBadge.ejs
- https://github.com/mdn/yari/blob/main/kumascript/macros/ExperimentalBadge.ejs
- https://github.com/mdn/yari/blob/main/kumascript/macros/NonStandardBadge.ejs
- https://github.com/mdn/yari/blob/main/kumascript/macros/ObsoleteBadge.ejs

There may be even more that I am not aware of.

## Conclusion

It seems that these icons are causing more problems than they are solving. As a result, this discussion is intended to come to a conclusion between readers, authors, and engineering as to how this information is best surfaced. Perhaps only showing these in the BCD tables using consistent iconography and colors? Perhaps a completely new approach is needed. Whichever solution is decided upon, we need to ensure that it is accessible to everyone.

**Comment by teoli2003** [comment]
We are trying to solve three goals here: conveying information, bringing more coherence to MDN, and better maintainability of the content.

1. **Conveying information** about the feature
  We want to tell the user to pay attention to this feature in five different places:
  - _On the feature reference page_. This is usually done using a page-wide banner at the top of the page (out of the scope of this problem)
![Capture d’écran 2022-05-11 à 16 08 53](https://user-images.githubusercontent.com/1466293/167869914-34aac2d8-4bb3-462e-8d3c-e7f685c8c6c0.png)

  - _In the sidebar_. Here we have a list of all related features, and we want to allow the users to navigate easily it. Here knowing that a feature is deprecated or non-standard, they hopefully want to try first standard features, and only fall back to iconized features if they can't solve their problem. Currently, we use colored icons in front of the link.
![Capture d’écran 2022-05-11 à 16 11 35](https://user-images.githubusercontent.com/1466293/167870593-82833730-920d-499b-a22b-e34733b8107f.png)

  - _In the definition list of the interface_. Similarly to the sidebars, we want to draw attention to the problem when the reader scans the definition list quickly.
 
![Capture d’écran 2022-05-11 à 16 22 14](https://user-images.githubusercontent.com/1466293/167872846-03630e9a-2770-4a12-8200-649ce1e8726a.png)

  -  _In the prose_. When reading some prose, especially prose that has not been updated recently and is outdated, we want to drag their (and ours) attention that what is said is outdated. We don't do this now, but having the icon next to the link would achieve this. (Ideally, this would be detected by Yari when generating the site, without a need of modifying the content or relying on a macro!) For example, all the links there lead to deprecated (and removed) features:
 
![Capture d’écran 2022-05-11 à 16 26 00](https://user-images.githubusercontent.com/1466293/167873772-40bb6813-3e80-412c-b3d1-9df529ec0542.png)


  -  _In the browser compat table_. That's, in my opinion, the neater:
 
![Capture d’écran 2022-05-11 à 16 24 37](https://user-images.githubusercontent.com/1466293/167873340-46c880dc-2171-4aa6-bc19-27bb1b036c4d.png)

2. **Coherence** in the way we annotate features
  - When using a website that is coherent, the reader feels at ease: they know the meaning of the symbol, they are used to the wording and they know where to find the information they are looking for quickly.
  - We want to have marks that mean the same thing, that have the same visual aspects (type of icon, colors), and that will always be where they should be. We don't have this right now.

3. **Maintenance** of data
  - The source of trust for this data is mdn/browser-compat-data. We should update the values there, and, automatically, after a week or less, the icons are updated on MDN.
  - We have two other ways to store this data on MDN. They predate mdn/browser-compat-data and **must** go away:
    - Sidebars read the _Tags:_ in the Yaml of each page and then apply the correct icons.
    - Definition lists have hardcoded (and unmaintainable) macros.
  
So I do think, we should unify all this. This is an intricate project and in order to succeed, feature creep should be avoided at all costs: for example, merging this with a redesign of the sidebar, even if this is needed, will likely make this project too big and fail.

I do like the way the icons are displayed in the browser compat table, even for multi-line entries. I think we should unify this: same colors and behaviors (on hover for example), immediately right of the term (and not left like in the sidebar), and the nice way they handle multiline entries:
![Capture d’écran 2022-05-11 à 16 07 34](https://user-images.githubusercontent.com/1466293/167869701-5f37cd48-bfc8-4e30-a748-da5f2dbb4041.png)
Maybe something similar can be done for sidebars!

I would be open to another solution than displaying icons for displaying this info, but I don't have any. We tried labels before, and it didn't work.

**Reply by wbamberg** [reply]
Great comment @teoli2003 , you saved me from trying to organize my thoughts! Especially this:

> The source of trust for this data is mdn/browser-compat-data. We should update the values there, and, automatically, after a week or less, the icons are updated on MDN.
>
> We have two other ways to store this data on MDN. They predate mdn/browser-compat-data and must go away:
>    - Sidebars read the Tags: in the Yaml of each page and then apply the correct icons.
>    - Definition lists have hardcoded (and unmaintainable) macros.

I feel like the rest of it - deciding how to represent these statuses to readers - is a quite separate problem from how to maintain and update the data.

For the "banners" (indicators on the page itself) we are not so far from being able to do this, with the big caveat about what do we do for pages that don't have a corresponding BCD entry, like API overview pages (this is the same issue we currently face for specifications). For the "badges" (indicators attached to links to the page) it's trickier.

**Comment by hamishwillee** [comment]
Great summary. Also strongly agree that there needs to be a separation of data storage and presentation. Ideally just one place for it to be stored, which would be the BCD, with fallback to page metadata if no BCD entry exists.

Some minor asides
- Obsolete should no longer appear anywhere. The macros may exist but the concept got rolled into the deprecated catch all some years ago.
- Read-only for properties is also useful tagging. You might decide to have this as metadata too, and do icon iff the property is displayed as the definition in a definition list, sidebar etc.

**Reply by teoli2003** [reply]
I think _obsolete_ is really a thing of the past on mdn/content: 
- `rg "\{\{ *Obsolete" -i -l` leads to 0 entry (Yeah!). This has been confirmed by @queengooborg who marked them as deprecated in yari: they'll rise a flaw if somebody try to use them).
- `rg "\- Obsolete" -i -l` (detecting mostly tags) leads to 1 entry that is MDN meta-docs about what is deprecated what is obsolete.

It is a different story on mdn/translated-content.

The _read-only_ information should be read from a reference source, and added automatically in definition lists (as well as in the first sentence), and I think we should get the info form w3c/webref where it should be stored.
